File: twoSum.py
""" create a Hash table, nums -> dict
Example:
      Input: nums = [2, 7, 1, 5]
	  Return: table = {2:0, 7:1, 1:2, 5:3}
 """
def createHashTable(nums):
	table = {}
	for i in range(len(nums)):
		table[nums[i]] = i
	return table


""" Two-pass Hash table """
def twoSum(nums, target):
	table = createHashTable(nums)
	for i in range(len(nums)):
		if target - nums[i] in table:
			return [i, table[target-nums[i]]]
	return 


""" One-pass Hash table """
def twoSumOnePass(nums, target):
	table = createHashTable(nums)
	for i, num in enumerate(nums):
		if target - num in table:
			return [i, table[target-num]]
			break
		# table[num] = i
	return ([])



if __name__ == "__main__":
	nums = [2,5,1,7]
	result = twoSumOnePass(nums, 9)
	# result2 = twoSum(nums, 9)
	print(result)
	# print(result2)

File: twoSum.java
import java.util.HashMap;
import java.util.Map;

/** Given an array of integers, return indices of the two numbers
 *  such that they added up to a specific target.
 * Example:
 *  Given nums = [2, 7, 11, 15], target = 9
 *
 *  Because num[0] + num[1] = 9
 *  return [0, 1]
 *
 * */
public class twoSum {

    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        int[] x = new int[]{0, 0};
        for (int i = 0; i < nums.length; i++) {
            map.put(nums[i], i);
        }

        for (int i = 0; i < nums.length; i++) {
            int other = target - nums[i];
            if (map.containsKey(other)) {
                int pos = map.get(other);
                if (pos == i)  continue;
                x[0] = i;
                x[i] = pos;
                return x;
            }
        }
        throw new IllegalArgumentException("No two sum solution");
    }

}
